### PR TITLE
Big overhaul to load additional data from Salesforce

### PR DESCRIPTION
**Pull request by jacobdadams**
Now pulls in records from Salesforce to calculate five different reports that were previously done manually in local jupyter notebooks.

Refactor of skid template into a Skid class with an eye towards a future pattern of having a ugrc-skid module with a Skid class with unimplemented methods that the client script is responsible for defining, like forklift pallets. This should hopefully reduce boilerplate code in skid scripts and improve DRY.

Leaving as draft for now pending a meeting with Stevie to discuss the calculations, but the framework and logic are all here.

### CODE REVIEW

File: src/wmrc/main.py [hunk 1]
```diff
@@ -22,158 +22,170 @@
...
-        sendgrid_api_key (str): The API key for sendgrid for this particular application
+    def _initialize_supervisor(self):
+        """A helper method to set up logging and supervisor
 
-    Returns:
-        Supervisor: The supervisor object used for sending messages
-    """
+        Args:
+            log_path (Path): File path for the logfile to be written
+            sendgrid_api_key (str): The API key for sendgrid for this particular application
```
Comment (steveoh) [inline src/wmrc/main.py:78]: **chore**: synchronize the docs with the method definition

File: src/wmrc/main.py [hunk 2]
```diff
@@ -187,84 +199,424 @@ def process():
...
+        )
+        salesforce_extractor = extract.SalesforceRestLoader(self.secrets.SF_ORG, salesforce_credentials)
+
+        salesforce_records = helpers.SalesForceRecords(salesforce_extractor)
+        salesforce_records.extract_data_from_salesforce()
+
+        return salesforce_records
+
+
+class Summarize:
```
Comment (steveoh) [inline src/wmrc/main.py:404]: **question**: if every method is static, is the class used for organization? Would it be simpler to import a summarize.py file?

**Review by steveoh — COMMENTED** [review]
This looks pretty nice. A lot of dataframe love!
My review is mostly questions. 
There are some commented out code that should be removed but i didn't call them out.
I thought you wanted to split the logic from the base class to help identify the abstraction. Did something worth sharing change your mind?

File: src/wmrc/helpers.py [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+    """
+
+    input_dataframe["x"] = 12_495_000
+    input_dataframe["y"] = 5_188_000
+
+    spatial_dataframe = pd.DataFrame.spatial.from_xy(input_dataframe, "x", "y", sr=4326)
+
+    spatial_dataframe.drop(columns=["x", "y"], inplace=True)
+
+    return spatial_dataframe
```
[the remark this answers is not in the document]
Comment (jacobdadams) [reply src/wmrc/helpers.py:33]: Yeah, I call it in three different update functions.

File: src/wmrc/helpers.py [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+        report runs to the Salesforce column names.
+
+        Raises:
+            ValueError: If a field from the manual reports is not found in the Salesforce columns
+        """
+
+        #: Get a dataframe of the first 200 rows with all the columns so we can extract the needed columns
+        df_for_columns = self.salesforce_extractor.get_records(
+            "services/data/v60.0/query/", "SELECT FIELDS(ALL) from Application_Report__c LIMIT 200"
+        )
```
[the remark this answers is not in the document]
Comment (jacobdadams) [reply src/wmrc/helpers.py:94]: 200 is the max you can get using FIELDS(ALL)... but you're right, we really only need one.

File: src/wmrc/main.py (hunk 2 above)
Comment (jacobdadams) [reply src/wmrc/main.py:404]: Good catch, that would be simpler.

File: src/wmrc/main.py [hunk 5]
```diff
@@ -22,158 +22,170 @@
...
 
-def _get_secrets():
-    """A helper method for loading secrets from either a GCF mount point or the local src/wmrc/secrets/secrets.json file
+class Skid:
+    def __init__(self):
+        self.secrets = SimpleNamespace(**self._get_secrets())
+        self.tempdir = TemporaryDirectory(ignore_cleanup_errors=True)
+        self.tempdir_path = Path(self.tempdir.name)
+        self.log_name = f'{config.LOG_FILE_NAME}_{datetime.now().strftime("%Y%m%d-%H%M%S")}.txt'
+        self.log_path = self.tempdir_path / self.log_name
```
[the remark this answers is not in the document]
Comment (jacobdadams) [reply src/wmrc/main.py:38]: The log gets sent as an attachment with the email... do you think that's overkill with it being logged in GCP?

File: src/wmrc/helpers.py [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+    """
+
+    def __init__(self, salesforce_extractor: palletjack.extract.SalesforceRestLoader):
+        self.salesforce_extractor = salesforce_extractor
+        self.field_mapping: Mapping[str, str] = {}
+        self.county_fields: list[str] = []
+
+        self._build_field_mapping()
+
+    def extract_data_from_salesforce(self):
```
[the remark this answers is not in the document]
Comment (jacobdadams) [reply src/wmrc/helpers.py:48]: Yeah, the field list needs to be built and added to the query string. This is specific to the wmrc skid, not part of a broader Skid module so I'm not worried about wider reuse.

File: src/wmrc/main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (steveoh) [reply src/wmrc/main.py:38]: I suppose that is up to you. Log monitoring could spot issues and save you an email? Whatever allows you the most visibility to keep things running smoothly.

**Review by stdavis — DISMISSED** [review]

File: src/wmrc/main.py [hunk 7]
```diff
@@ -22,158 +22,163 @@
...
 
-        skid_supervisor = _initialize(log_path, secrets.SENDGRID_API_KEY)
-        module_logger = logging.getLogger(config.SKID_NAME)
+    def process(self):
+        """The main method that does all the work."""
 
-        # module_logger.info(pprint.pformat(dict(os.environ)))
-        # module_logger.info(locale.getdefaultlocale())
-        # module_logger.info(locale.getlocale(locale.LC_NUMERIC))
+        #: Set up secrets, tempdir, supervisor, and logging
```
[the remark this answers is not in the document]
Comment (jacobdadams) [reply src/wmrc/main.py:137]: good catch, thanks.

**Review by stdavis — APPROVED** [review]

**Review by stdavis — APPROVED** [review]

**Review by stdavis — APPROVED** [review]

File: src/wmrc/config.py [hunk 8]
```diff
@@ -35,8 +35,8 @@
 
 COUNTIES_ITEMID = "90431cac2f9f49f4bcf1505419583753"
 
-FACILITIES_LAYER_ITEMID = "d6736e9525274eadb7461b2fd08c426d"  #: Beta version
-# FACILITIES_LAYER_ITEMID = "056bbc52ff3240f6b69666750a61aeff"  #: Live version
+# FACILITIES_LAYER_ITEMID = "d6736e9525274eadb7461b2fd08c426d"  #: Beta version
+FACILITIES_LAYER_ITEMID = "056bbc52ff3240f6b69666750a61aeff"  #: Live version
```
Comment (stdavis) [inline src/wmrc/config.py:39]: Would this be easier to maintain via an environment variable?

Comment (jacobdadams) [reply src/wmrc/config.py:39]: It might be, but I'd still need it locally for local runs so I'm just going to run with it for now.

**Pull request merged**